Feature: Added filtering ability to ncurses ui for node output streams

Cartoonman opened the pull request:
I added a feature that allows the user to filter the ros log outputs on the node(s) they select. This only affects the ncurses ui, and does not propagate to the logging mechanism; just the ui output. Useful for specifically trying to debug a particular node that crashed by observing it's output if other nodes are producing too much output to easily see. 

Usage: 
Select node as usual, and press f to filter. The foreground text color will become white to denote filtered node. 
Upon filtering the first node, all other nodes will be filtered out until the last filtered node is unfiltered. 
To unfilter, select filtered node and press f again.

Did my best to keep variable naming conventions and bracketing consistent with source file.

xqms commented:
Cool, this looks like a very useful feature! :+1:
I'm currently very busy but I will try to check it out next week ;-)

xqms commented:
@ros-pull-request-builder retest this please

Cartoonman commented:
I made some modifications in my second commit which I'll explain:

I changed the color indication for filtered nodes to be the left keybox of each node's background color. In testing I felt more comfortable using this approach to denote filtered nodes than the colored text (easier to read and better to notice). Matter of opinion, so feel free to change it if you feel theres a better way to highlight them.

I also added an exclusive filter option (x) that will exclude node(s) from the output. The inclusive and exclusive filters are orthogonal and do not mix; so if you wish to change from one to the other you must return to an unfiltered state first. This is because (at least in my setup) I have one node that output a lot of data, but the other nodes don't output as frequently, so removing the one node allows me to have a clearer picture of the system's performance. 

Inclusively-filtered nodes are highlighted with blue, where exclusively-filtered nodes are highlighted with red. I tested it to make sure it works with 256-color mode and truecolor mode on ubuntu 16.04 terminal.

I will be traveling in the next week so if you have questions I can answer them when I get back.

xqms commented on rosmon_core/src/ui.h:
```diff
@@ -37,10 +37,20 @@ class UI
 		Terminal::Parser parser;
 	};
 
+	enum FilterMode
```
Is there a reason why this is a bitmask? Can a node be both InclusiveFiltered and ExclusiveFiltered?
Otherwise, please change it to a usual enum.

Cartoonman replied:
Removed entirely, so no longer an issue.

Cartoonman commented:
Good question,

The inclusive/exclusive system from a functional standpoint was my attempt of optimizing the least amount of steps required to filter/unfilter node outputs for most user's usecases, such that the worst case number of 'manual filter keystrokes' occurs at 50% (e.g. it takes 1 option cycle to only display output from only 1 node, and 1 option cycle to display output for all but 1 node. Likewise it takes 2 option cycles to display outputs from only 2 nodes, and vice versa. The worse case would be n/2 option cycles to display/not display outputs from n/2 nodes where n = number of nodes in launch file.)

I could implement your suggestion, but assuming just a "mute", "mute all others", and "unmute all" set of options would lead you to a worse case n-1 situation if a user wanted to see just two node's output (I'm not sure how to use 'mute all others' twice without barging against the semantics of 'all others').

To your point, I was thinking of a 'mute/unmute all' option, but I was worried that adding 3 totally new options would risk make the option list too long to be usable on small terminal screens. A workaround to this is to add a derivative option menu. (e.g.` f - filter options`, and when selected, displays the given filter options for the node, thereby compressing the options list (and allowing for n-character options, like `fm`, similar to how Vim does it)

Additionally, I can see how having two states makes it confusing (inclusive and exclusive), and I think having a single visible state (Muted) is a simpler indicator for the user on the state of the filtering. Perhaps I can start off with the following:

```
Mute
Mute all
Unmute
Unmute all
```

With these choices I think it's a decent compromise to the optimization problem mentioned earlier, and should be clearer to the user on what is going on and how to manipulate the filtering rules. What do you think?

xqms commented:
> I could implement your suggestion, but assuming just a "mute", "mute all others", and "unmute all" set of options would lead you to a worse case n-1 situation if a user wanted to see just two node's output (I'm not sure how to use 'mute all others' twice without barging against the semantics of 'all others').

In that situation, I would select the first node, "mute all others", then select the second node, and unmute it.

> but I was worried that adding 3 totally new options would risk make the option list too long to be usable on small terminal screens.

I would propose "Mute all" and "Unmute all" as global shortcuts, i.e. they would be displayed when no node is selected. We would probably use one of the function keys (e.g. F9/F10) for this.

We should definitely display a warning if muting is active. This kind of feature is dangerous if the user activates it by accident...

Cartoonman commented:
I would say if we have these two global functions implemented, then 'Mute' and "Unmute" will be more than sufficient for the user to further modify the filtering configuration. Keeps it simple. 

My idea of letting the user know was to leverage the new uniform state of 'muted' by colortagging all muted nodes to one particular color (by coloring the key box, not the node name box). If you know of a different or more noticeable way of letting the user know that some nodes are muted, let me know. I'll start working on modifying the filtering system to the design described above.

xqms commented:
> I would say if we have "Mute all" and "Unmute all" as two global functions implemented, then 'Mute' and "Unmute" will be more than sufficient for the user to further modify the filtering configuration. Keeps it simple.

:+1:

> My idea of letting the user know was to leverage the new uniform state of 'muted' by colortagging all muted nodes to one particular color (by coloring the key box, not the node name box). If you know of a different or more noticeable way of letting the user know that some nodes are muted, let me know.

That sounds good. In addition, I would display something like "Caution: some nodes are muted." in the status line (where the shortcut keys are usually displayed).

Cartoonman commented:
@xqms All of the changes should be ready for review.

xqms commented:
Cool, thanks! I like the overall design :+1:

I hacked together some code in the `Terminal` class to be able to support function keys and changed to F9/F10 for muting/unmuting globally. Those keys have a more "global" feel to it ;-)

Regarding the implementation, I would like an `std::vector<bool>` better than an `std::set<std::string>`. The list of nodes never changes, so we can use the index as key and avoid the string lookup.

Cartoonman commented:
I had some reservations using the Function keys because there are cases where the function keys are either binded by an overlay (e.g. VNC into computer), or don't cleanly propagate (e.g. in a VM), so in such cases the function keys may not work. What's the value in having an optional flag to fallback to +/- keys if the function keys aren't available for the user?

On the point regarding `std::vector<bool>`, I didn't see a way to know which node was given to the `void UI::log(const std::string& channel, const std::string& str)` function besides comparing the `channel` string to the list of nodes given by `m_monitor->nodes()`, which is where I went with the std::set<std::string> route. I would have preferred std::unordered_set since it implements a hashmap in the backend so lookups are constant time, but that's a C++11 feature. A `std::vector<bool>` would work if I had some way of knowing the given channel's node index in the log function without having to do a string lookup.

xqms commented:
> I had some reservations using the Function keys because there are cases where the function keys are either binded by an overlay (e.g. VNC into computer), or don't cleanly propagate (e.g. in a VM), so in such cases the function keys may not work. What's the value in having an optional flag to fallback to +/- keys if the function keys aren't available for the user?

Do you actually have such a use case? In my experience, VNC + VMs propagate such keys just fine.
In any case, my approach would be to wait until someone reports an issue with the keybinding.
We probably should implement a proper keybinding configuration file in that case.

> On the point regarding `std::vector<bool>`, I didn't see a way to know which node was given to the `void UI::log(const std::string& channel, const std::string& str)` function

Ah okay, I didn't think about that -.- Yep, in that case a string lookup is probably the best we can do (without touching the log signal infrastructure).

>  I would have preferred std::unordered_set since it implements a hashmap in the backend so lookups are constant time, but that's a C++11 feature.

Feel free to do that - rosmon is using C++11 everywhere :-)

Cartoonman commented:
I tested out the function keys and F9 and F10 seem fine on both my VNC server and my VM. I know F1 and F2 do different things for my VNC server (and they don't propagate without likely some special shortcut command). I agree with you; wait till someone complains :)

Also I was being dumb; I just didn't include unordered_set. Thats why it wasn't working. I replaced it now and functionality works.

xqms commented:
Awesome, thanks :+1:

I'll test this tomorrow on some of our "production" setups and will merge if everything works :-)

xqms commented:
Everything seems to work fine, so I'll go ahead and merge into master :-)

Thank you for the contribution!
